src/daft-core/src/series/ops/utf8.rs

```diff
@@ -332,4 +332,59 @@ impl Series {
             ))),
         }
     }
+
+    pub fn utf8_substr(&self, start: &Series, length: &Series) -> DaftResult<Series> {
+        if self.data_type() == &DataType::Null {
+            return Ok(self.clone());
+        }
+        if self.data_type() != &DataType::Utf8 {
+            return Err(DaftError::TypeError(format!(
+                "Substr not implemented for type {}",
+                self.data_type()
+            )));
+        }
+
+        // This part is only illustrative for now
+        match start.data_type() {
```

Comment:
The data type mapping part is a bit awkward, I could use some pointers here on how to proceed.

Theoretically, `start.data_type()` and `length.data_type()` could be of different data types. That occurs especially frequently when both of them are expressions. But firstly, creating a matrix of all possible type combinations here like this is a bit of a pain, and secondly, if they are indeed of a different data types, signature of the computation function needs to be changed to something like this:

```
    pub fn substr<I, U>(&self, start: &DataArray<I>, length: &DataArray<U>) -> DaftResult<Utf8Array>
    where
        I: DaftIntegerType,
        <I as DaftNumericType>::Native: Ord,
        U: DaftIntegerType,
        <U as DaftNumericType>::Native: Ord,
```
But to generalize variables like this I will need to update some helper functions e.g. `parse_inputs` as well.
I am fairly new to Rust, so maybe someone can recommend some clever solution here.

Reply:
This is a good question! I actually realized we have a lot of boilerplate in the string kernel code and just fixed it up with #2276. You can check out the `with_match_integer_daft_types` kernel, which will eliminate the need for you to do that matching matrix.

In terms of the array `substr` function signature, that looks fine! You would need two generic types in case `start` and `length` are different.

Reply:
Oh, that is much nicer, I will take a look. 
Thanks!